### PR TITLE
Streaming message handling: DRY, improve type safety

### PR DESCRIPTION
**Pull request by lguychard**
- New `messageHandlers` record guarantees that all messages are handled at compile time
- Avoids verbose duplication of `subscription.add(...)` blocks for each new message type
- Removes the possibility of event name typos (`fromEvent(eventSource, 'fileamtches')`)
- Removes the possibility of mapping events to an inappropriate type, which would've been easy to do with the previous approach, eg:

```typescript
        subscriptions.add(	
            observeMessages<Alert>(eventSource, 'filematches')	
                .pipe(map(alert => ({ type: 'alert' as const, data: alert })))	
                .subscribe(observer)	
        )
```

### CODE REVIEW

File: client/web/src/search/stream.tsx [hunk 1]
```diff
@@ -370,20 +370,50 @@ export const switchToGQLISearchResults: OperatorFunction<SearchEvent, GQL.ISearc
...
+                    throw new Error(`Could not parse ${type} message data in streaming search`)
+                }
+            }),
+            map(data => ({ type, data } as T))
+        )
+        .subscribe(observer)
+
+const messageHandlers: Record<
+    SearchEvent['type'],
+    (type: SearchEvent['type'], eventSource: EventSource, observer: Subscriber<SearchEvent>) => Subscription
```
Comment (limitedmage) [inline client/web/src/search/stream.tsx:397]: The only issue I see here from a type checking perspective is that the `SearchEvent['type']` in the first line does not have to match the `SearchEvent['type']` in the second line. I tried for a while last week to fix it but couldn't find a way to strongly type this unfortunately :(

**Review by limitedmage — APPROVED** [review]

File: client/web/src/search/stream.tsx (hunk 1 above)
Comment (lguychard) [reply client/web/src/search/stream.tsx]: Oh I didn't think of enforcing that! Here goes: https://github.com/sourcegraph/sourcegraph/pull/15522/commits/117310dca8e6c7ac8b7d8b0ff99bd97e9e8a7a2e

Comment (lguychard) [reply client/web/src/search/stream.tsx]: `type` is now correctly inferred in the callback :)

image

**Review by keegancsmith — APPROVED** [review]

**Comment by keegancsmith** [comment]
merged the base into master, then merged master into this branch :) If CI is happy I assume we are good to go for merging this in.

**Pull request merged**